src/simcams/simulated.camera.c

```diff
@@ -429,27 +437,32 @@ simcam_start(struct Camera* camera)
 }
 
 static enum DeviceStatusCode
-simcam_stop(struct Camera* camera)
+simcam_execute_trigger(struct Camera* camera)
 {
     struct SimulatedCamera* self =
       containerof(camera, struct SimulatedCamera, camera);
-    self->streamer.is_running = 0;
-    event_notify_all(&self->software_trigger.event);
-    condition_variable_notify_all(&self->im.frame_ready);
 
-    TRACE("SIMULATED CAMERA: thread join");
-    ECHO(thread_join(&self->streamer.thread));
+    lock_acquire(&self->im.lock);
+    self->software_trigger.triggered = 1;
+    condition_variable_notify_all(&self->software_trigger.trigger_ready);
+    lock_release(&self->im.lock);
 
-    TRACE("SIMULATED CAMERA: exiting");
     return Device_Ok;
 }
 
 static enum DeviceStatusCode
-simcam_execute_trigger(struct Camera* camera)
+simcam_stop(struct Camera* camera)
 {
     struct SimulatedCamera* self =
       containerof(camera, struct SimulatedCamera, camera);
-    event_notify_all(&self->software_trigger.event);
+    self->streamer.is_running = 0;
+    simcam_execute_trigger(camera);
+    condition_variable_notify_all(&self->im.frame_ready);
```

Comment:
Is `&self->im.lock` held by the caller on this notify? Does it need to be?

Generally, I think I usually signal/notify with the associated lock held, then immediately release it, as you do in `simcam_execute_trigger`. That also seems to be [recommended by the pthreads docs if you want predictable scheduling](https://pubs.opengroup.org/onlinepubs/007904975/functions/pthread_cond_broadcast.html), though I'm still trying to think through what means and if it matters in this case. Not sure about Windows.

Reply:
`&self->im.lock` is not held by this caller. It doesn't need to be. 

The idea is that `is_running` is really only ever written here. So we can avoid a lock and still be consistent. 

That said `simcam_execute_trigger` acquires the lock anyway...

Reply:
I wasn't really thinking about `is_running` and agreed you don't need a lock for that. But I am still thinking about the `frame_ready` condition variable.

Why does `frame_ready` get notified on stop? Is this forcing some cleanup somewhere? It could be worth adding an explanatory comment in the code if true. I might also expect `frame_id` to be updated before notifying as in `simulated_camera_streamer_thread`

Also note that `frame_ready` may also be notified (in `simulated_camera_streamer_thread`) as a result of notifying `trigger_ready` in `simcam_execute_trigger`. I guess the potential double notification shouldn't cause any serious issues, but it does make this concurrency scheduling harder to think about.

Reply:
`frame_ready` should really be called something like `frame_ready_or_abort`. 

It's used to block the calling thread in `simcam_get_frame` ([here](https://github.com/acquire-project/acquire-driver-common/pull/34/files#diff-b8b472097eb37f10cdde869a1b8b4fc87d791ffcf26f61bb5f1f5ebbb9eea867R483-R491)). You can see how `is_running` gets checked as soon as the lock gets reacquired after the wait. `frame_ready` gets notified because new frames aren't coming and we want the caller to continue.

Reply:
Thanks for clarifying the intent. I think this is fine. But this conversation did make me realize that I need to better understand the key threads in acquire.